app/classifier/frame-annotator.cjsx

```diff
@@ -0,0 +1,148 @@
+    right += pageXOffset
+    top += pageYOffset
+    bottom += pageYOffset
+    @setState sizeRect: {left, right, top, bottom, width, height}
+
+  getScale: ->
+    horizontal = @state.sizeRect?.width / @props.naturalWidth || 0
+    vertical = @state.sizeRect?.height / @props.naturalHeight || 0
+    {horizontal, vertical}
+
+  getEventOffset: (e) ->
+    scale = @getScale()
+    console?.log 'Subject scale is', JSON.stringify scale
+    x = (e.pageX - @state.sizeRect?.left) / scale.horizontal || 0
+    y = (e.pageY - @state.sizeRect?.top) / scale.vertical || 0
+    {x, y}
+
+  render: ->
+    taskDescription = @props.workflow.tasks[@props.annotation?.task]
+    TaskComponent = tasks[taskDescription?.type]
+    {type, format, src} = getSubjectLocation @props.subject, @props.frame
+
+    svgStyle = {}
+    if type is 'image'
+      # Images are rendered again within the SVG itself.
+      # When cropped right next to the edge of the image,
+      # the original tag can show through, so fill the SVG to cover it.
+      svgStyle.background = 'black'
+
+    svgProps = {}
+
+    if TaskComponent?
+      {BeforeSubject, InsideSubject, AfterSubject} = TaskComponent
+      hookProps =
+        workflow: @props.workflow
+        task: taskDescription
+        classification: @props.classification
+        annotation: @props.annotation
+        frame: @props.frame
+        scale: @getScale()
+        naturalWidth: @props.naturalWidth
+        naturalHeight: @props.naturalHeight
+        containerRect: @state.sizeRect
+        getEventOffset: this.getEventOffset
+
+      for task, Component of tasks when Component.getSVGProps?
+        for key, value of Component.getSVGProps hookProps
+          svgProps[key] = value
+
+    <div className="frame-annotator">
+      <div className="subject-area">
+        {if BeforeSubject?
+          <BeforeSubject {...hookProps} />}
+
+        <svg style=svgStyle viewBox="0 0 #{@props.naturalWidth} #{@props.naturalHeight}" {...svgProps}>
+          <rect ref="sizeRect" width={@props.naturalWidth} height={@props.naturalHeight} fill="rgba(0, 0, 0, 0.01)" fillOpacity="0.01" stroke="none" />
+
+          {if type is 'image'
+            <SVGImage src={src} width={@props.naturalWidth} height={@props.naturalHeight} />}
+
+          {if InsideSubject?
+            <InsideSubject {...hookProps} />}
+
+          {for anyTaskName, {PersistInsideSubject} of tasks when PersistInsideSubject?
+            <PersistInsideSubject key={anyTaskName} {...hookProps} />}
+        </svg>
+        {@props.children}
+
+        {if AfterSubject?
+          <AfterSubject {...hookProps} />}
+      </div>
+    </div>
+
+  handleChildLoad: (e) ->
+    @props.onLoad? e, @props.frame
+    {naturalWidth, naturalHeight} = e.target
+    unless @props.naturalWidth is naturalWidth and @props.naturalHeight is naturalHeight
+      @setState {naturalWidth, naturalHeight}, =>
+        @updateSize()
+
+  handleFrameChange: (frame) ->
+    @setState {frame}
```

Comment:
Should frame changes be handled at a higher level, by a component which has access to the whole subject? I'm confused here, because you have an immutable frame number via `props.frame` but also have a frame number stored in `state.frame`, which can be changed. Which frame should this component display?